### PR TITLE
feat: Add role assign related mails

### PR DESCRIPTION
**Pull request by howar31**
# Issue
Add role assign related mails:
- send_role_explorer
- send_role_actiontaker
- send_role_trailblazer
- send_role_downgrade (only route, for cronjob implemented in later task)

# Notice

# Dependency

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204934606026369

### CODE REVIEW

File: controllers/account.go [hunk 1]
```diff
@@ -124,6 +125,7 @@ func (mc *MembershipController) AuthByEmail(c *gin.Context, sendMailRoutePath st
 			}
 		}
 
+		go mc.sendAssignRoleMail(constants.RoleExplorer, email)
```
Comment (Aylie-Chou) [inline controllers/account.go:128]: Question: do we need to assign roles for the user before sending email ?

File: controllers/mail.go [hunk 2]
```diff
@@ -168,6 +173,138 @@ func (contrl *MailController) SendDonationSuccessMail(c *gin.Context) (int, gin.
 	return http.StatusNoContent, gin.H{}, nil
 }
 
+func (contrl *MailController) SendRoleExplorerMail(c *gin.Context) (int, gin.H, error) {
```
Comment (Aylie-Chou) [inline controllers/mail.go:176]: It seems that the four added functions have some common codes. Could we combine them into a function to reduce duplicate codes ?

File: template/role-trailblazer.tmpl [hunk 3]
```diff
@@ -0,0 +1,822 @@
...
+
+<p dir="ltr"><span style="color:#404040"><span style="font-size:16px">★ 透過電子報了解報導幕後和我們的工作內容</span></span></p>
+
+<p dir="ltr"><span style="color:#404040"><span style="font-size:16px">★ 獲得每季基金會工作報告，了解您的捐款對《報導者》與台灣社會的價值</span></span></p>
+
+<p dir="ltr"><span style="color:#404040"><span style="font-size:16px">★ 免費參與《報導者》付費活動</span></span></p>
+
+<p dir="ltr"><span style="color:#404040"><span style="font-size:16px">★ 享有《報導者》新書出版品 6 折優惠</span></span></p>
+
+<p dir="ltr"><span style="color:#404040"><span style="font-size:16px">★ 同時成為國家兩廳院的聯名會員</span></span></p>
```
Comment (Aylie-Chou) [inline template/role-trailblazer.tmpl:623]: It seems the content of html template is a bit different with the provided doc on [dirve](https://drive.google.com/drive/folders/1T1gDVHoiuB67XdgZBkzm63nyN7oQ49Bd).
Please help to check it.

**Review by Aylie-Chou — CHANGES_REQUESTED** [review]
Mostly looking good, and there's one thing to check:
would we need to add feature toggle here since `go-api` repo do not have `dev` branch/environment ?

File: template/role-trailblazer.tmpl (hunk 3 above)
Comment (howar31) [reply template/role-trailblazer.tmpl:623]: the template was updated on 20th July after this commit :/

File: controllers/account.go (hunk 1 above)
Comment (howar31) [reply controllers/account.go:128]: The role has been assigned in `InsertUserByReporterAccount` before sending the mail.

File: controllers/mail.go (hunk 2 above)
Comment (howar31) [reply controllers/mail.go:176]: done

**Review by Aylie-Chou — CHANGES_REQUESTED** [review]
Previous comments all resolved !

As the discussion in refinement meeting, please help to check following feature:
- If user already on that role, we would not send upgrading mail again
- sending explorer email only when user is `activated`

Feel free for any questions/discussions !

File: controllers/account.go [hunk 4]
```diff
@@ -106,6 +107,7 @@ func (mc *MembershipController) AuthByEmail(c *gin.Context, sendMailRoutePath st
 
 		// try to find record by email in users table
 		matchedUser, err = mc.Storage.GetUserByEmail(email)
+		roleCheck, _ := mc.Storage.HasRole(matchedUser, constants.RoleExplorer)
```
Comment (Aylie-Chou) [inline controllers/account.go:110]: Question: how would we handle the error of `HasRole` ?

File: controllers/donation.go [hunk 5]
```diff
@@ -777,10 +818,15 @@ func (mc *MembershipController) CreateADonationOfAUser(c *gin.Context) (int, gin
 		}
 
 		// Call AssignRoleToUser to assign role to user
+		roleCheck, _ := mc.Storage.HasRole(matchedUser, constants.RoleActionTaker)
```
Comment (Aylie-Chou) [inline controllers/donation.go:821]: Question: if the user has already been `trailblazer`, would the role changed to `action taker` here ?

**Review by Aylie-Chou — CHANGES_REQUESTED** [review]
Please have a look at comments & would we add `feature toggle` in this pr ?

File: controllers/donation.go (hunk 5 above)
Comment (howar31) [reply controllers/donation.go:821]: Yes, it will.  The `HasRole` only check for specific role (`Action Taker` here, for example).

Comment (Aylie-Chou) [reply controllers/donation.go:821]: Is the role downgrading the expected spec ?

File: controllers/account.go (hunk 4 above)
Comment (Aylie-Chou) [reply controllers/account.go]: it sounds good.
could we also add the same handling in other `HasRole` check place (`controller/donation`) ?

Comment (howar31) [reply controllers/account.go]: All `HasRole` has similar error handlings.  They won't effect the original logics.

File: controllers/donation.go (hunk 5 above)
Comment (howar31) [reply controllers/donation.go]: Fix patched

**Comment by howar31** [comment]
According to discussion in meeting, the feature toggle will be implemented in another PR

File: controllers/donation.go (hunk 5 above)
Comment (howar31) [reply controllers/donation.go]: orz

**Pull request merged**